### PR TITLE
Remove test retry plugin-specific configuration cache issues

### PR DESCRIPTION
**Pull request by ghale**
See https://github.com/gradle/test-retry-gradle-plugin/issues/70.

This removes the issues related to configuration cache that can be fixed in the retry plugin itself.  There are still a few issues that need to be resolved on the Gradle Build Tool side before the retry plugin is fully compatible with the configuration cache:

- Make Groovy plugin compatible with config cache (https://github.com/gradle/gradle/issues/13460)
- Remove project object access from TestNGTestFramework constructor (https://github.com/gradle/gradle/issues/14287)

With these changes, projects that use the `java` plugin and use either Junit4 or Junit5 will be able to use the configuration cache with the retry plugin.  Projects that use TestNG or the `groovy` plugin will not be compatible until the issues above are fixed.

### CODE REVIEW

File: plugin/src/test/groovy/org/gradle/testretry/testframework/TestNGConfigCacheFuncTest.groovy [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.testretry.testframework
+
+import org.gradle.testretry.AbstractConfigCacheFuncTest
+import spock.lang.Ignore
+import spock.lang.Issue
+
+@Ignore
```
[the remark this answers is not in the document]
Comment (ghale) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/TestNGConfigCacheFuncTest.groovy:22]: I couldn't figure out how to use that in this situation because `PendingFeature` is only applicable to methods, not classes.  Since the methods are (by design) in the parent class, that wouldn't work.

Another option would be to build a subclass-specific check into the `shouldTest` so that we automatically start checking this after 6.7 (when my fix should be available).  I'll do that unless you have an idea how to make `PendingFeature` work.

File: plugin/src/test/groovy/org/gradle/testretry/AbstractConfigCacheFuncTest.groovy [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        assertConfigurationCacheIsReused(result, gradleVersion)
+
+        where:
+        gradleVersion << GRADLE_VERSIONS_UNDER_TEST
+    }
+
+    void shouldTest(String gradleVersion) {
+        // Configuration cache is supported after 6.6
+        Assume.assumeTrue("$gradleVersion does not support configuration cache", VersionNumber.parse(gradleVersion).with {
+            it.major > 6 || (it.major == 6 && it.minor >= 1)
```
[the remark this answers is not in the document]
Comment (ghale) [reply plugin/src/test/groovy/org/gradle/testretry/AbstractConfigCacheFuncTest.groovy:84]: Well, it was "instant execution" starting in 6.1, then in 6.5 it was changed to configuration cache but with a slightly different argument, then in 6.6 it is mostly stabilized and the validation was turned on.  I originally had this test only checking for 6.6+ but then changed my mind and decided it was worth checking in earlier versions, too, but forgot to change the comment.

I lifted the condition straight out of the old `InstantExecutionFuncTest`, but I'll change it to use `GradleVersion`.

File: plugin/src/test/groovy/org/gradle/testretry/testframework/TestNGConfigCacheFuncTest.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonnybbb) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/TestNGConfigCacheFuncTest.groovy]: Agree with your proposed solution 👍

**Review by jonnybbb — APPROVED** [review]
LGTM

File: plugin/src/test/groovy/org/gradle/testretry/AbstractConfigCacheFuncTest.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonnybbb) [reply plugin/src/test/groovy/org/gradle/testretry/AbstractConfigCacheFuncTest.groovy]: thanks for the details - 👍

**Comment by ghale** [comment]
Ok, so it turns out there is a relatively painless workaround to get the groovy plugin working with config cache (add groovy implementation jar to implementation/testImplementation classpath).  So, I've also added a Spock config cache test as well.  

https://github.com/gradle/gradle/issues/14287 has been fixed in the latest 6.7 nightly, too, so config cache should be functional with test-retry plugin in at least the most common cases (TestNG only functional with 6.7).  We can address any edge cases as they are discovered.

@jonnybbb I think this is in a good state now.  Let me know if you find anything else.

**Comment by etiennestuder** [comment]
@ghale some questions:
- which groovy implementation specifically, groovy-all or ...?
- can you explain why this workaround is needed / fixes the current issue?
- so one would add the groovy implementation as `provided` to ensure it does not end up being available at runtime?

**Comment by ghale** [comment]
> which groovy implementation specifically, groovy-all or ...?

groovy-all is known to work, but I think it could be any groovy variant that works with the Groovy plugin: https://docs.gradle.org/current/userguide/groovy_plugin.html#header.

> can you explain why this workaround is needed / fixes the current issue?

Ok, so some of this was my own misunderstanding.  I had originally used some of the stuff from the base functional test class which applies the `groovy` plugin to all projects (even if they are testing with java-only tests).  The idea is that `groovy` also applies the `java` plugin, so all tests get what they need, even if there is a little more than they need.  I was originally testing only java tests (TestNG) but using the shared logic from the base class that was applying the `groovy` plugin and getting a failure with config cache.  When I changed it to just the `java` plugin, it was working.  That, combined with the high level documentation [here](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:plugins) led me to the conclusion that this was a more general issue with groovy support.  Turns out, I was only running into https://github.com/gradle/configuration-cache/issues/270 because there weren't any groovy sources in my java-only test project and with config cache enabled, it fails to find the groovy runtime even though it doesn't actually need it.  In fact, it would work fine with a "normal" groovy project because those would (by nature) have the groovy implementation available on the implementation/testImplementation configuration.  I hope that makes sense.

> so one would add the groovy implementation as provided to ensure it does not end up being available at runtime?

If you have a pure groovy project, it should just work.  If you have a mixed java project with groovy test classes, you'll need to add the groovy runtime to the compileOnly (or implementation) configuration to work with config cache.  If you have a pure java project with the groovy plugin applied, you can either just switch to the java plugin, or add the groovy runtime to the compileOnly/testCompileOnly (or implementation/testImplementation) configurations.  It's not ideal, but it's a pretty low effort workaround until https://github.com/gradle/configuration-cache/issues/270 is fixed.

File: plugin/src/test/groovy/org/gradle/testretry/AbstractConfigCacheFuncTest.groovy [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+package org.gradle.testretry
+
+import org.gradle.testkit.runner.BuildResult
+import org.gradle.testkit.runner.GradleRunner
+import org.gradle.util.GradleVersion
+import org.junit.Assume
+import org.spockframework.util.VersionNumber
+import spock.lang.Unroll
+
+abstract class AbstractConfigCacheFuncTest extends AbstractGeneralPluginFuncTest {
```
Comment (ldaley) [inline plugin/src/test/groovy/org/gradle/testretry/AbstractConfigCacheFuncTest.groovy:25]: We don't need a separate hierarchy for this. 

Please find a way to incorporate this into the existing hierarchy. You might need to add a new base class for the framework specific tests.

**Review by ldaley — CHANGES_REQUESTED** [review]

File: plugin/src/test/groovy/org/gradle/testretry/AbstractConfigCacheFuncTest.groovy (hunk 3 above)
Comment (ghale) [reply plugin/src/test/groovy/org/gradle/testretry/AbstractConfigCacheFuncTest.groovy]: Good point.  I think I expected it to diverge more when I started this.  I've merged these classes with the existing functional test classes and created a base framework class.  Let me know if you think we need to do more here.

**Comment by etiennestuder** [comment]
Thanks for the helpful explanation, @ghale. I applied the workaround and it worked.

**Review by ldaley — APPROVED** [review]
